examples/equilibria/coilset_opt_problem_tutorial.ex.py

```diff
@@ -0,0 +1,434 @@
+x_hfs = np.array([1.42, 1.06, 0.81, 0.67, 0.62, 0.62, 0.64, 0.60])
+z_hfs = np.array([4.80, 5.09, 5.38, 5.72, 6.01, 6.65, 6.82, 7.34])
+
+x_legs = np.concatenate([x_lfs, x_lfs, x_hfs, x_hfs])
+z_legs = np.concatenate([z_lfs, -z_lfs, z_hfs, -z_hfs])
+
+legs_isoflux = IsofluxConstraint(
+    x_legs, z_legs, ref_x=x_lcfs[2], ref_z=z_lcfs[2], constraint_value=0.1, tolerance=0.5
+)
+
+magnetic_targets = MagneticConstraintSet([lcfs_isoflux, legs_isoflux])
+
+# %% [markdown]
+
+# ## Optimisation Algorithm, Conditions and Parameters
+#
+# We next define the Optimiser: the optimiser algorithm, conditions and
+# parameters to be used. There is no one-size-fits-all approach here,
+# as the best optimiser for a given problem will depend strongly on the
+# optimisation objective and constraints (if any) being applied in the
+# problem.
+#
+# `Optimiser` is currently only a wrapper for NLOpt based optimisers, and as
+# such, the range of algorithms that may be chosen is determined by those
+# available in the NLOpt library. Not all algorithms will work for all
+# objectives - some require gradient information, for example - and
+# some require additional parameters. Check the NLOpt API documentation
+# for more details.
+#
+# Care must also be taken to ensure the termination criteria for the
+# optimisation are suitable. Otherwise, the optimisation may never
+# stop running if tolerances are too tight, or may stop early and
+# return poorly optimised states if the maximum number of evaluations
+# is too low or tolerances are too large.
+#
+# **Below we will directly input the optimisation algorithm, conditions and**
+# **parameters when defining the `TikhonovCurrentCOP`**
+
+# ## Iterators
+
+# The `CoilsetOptimisationProblem` is only used to optimise the coilset state
+# at fixed plasma psi;
+# the Grad-Shafranov equation for the plasma is not guaranteed to still be
+# satisfied
+# after the coilset state is optimised.
+#
+# Picard iteration is therefore used to find a self-consistent solution,
+# employed
+# by `Iterator` objects. `PicardCoilsetIterator` specifies a scheme in which a
+# Grad-Shafranov iteration is used to update the plasma psi alternate with
+# Coilset
+# optimisation at fixed plasma psi until the psi converges for the `Equilbrium`.
+#
+# ## Unconstrained Optimisation
+#
+# However, a poor initial `Equilibrium` (with corresponding `coilset`) will
+# lead
+# to difficulties during the Picard iteration used to find a self-consistent
+# solution.
+# It is therefore useful to perform a coarser fast pre-optimisation to try
+# find a
+# self-consistent state within the basin of convergence of the
+# full constrained `OptimisationProblem`.
+
+# %%
+
+unconstrained_cop = UnconstrainedTikhonovCurrentGradientCOP(
+    coilset, eq, magnetic_targets, gamma=1e-8
+)
+unconstrained_iterator = PicardIterator(
+    eq,
+    unconstrained_cop,
+    fixed_coils=True,
+    plot=False,
+    relaxation=0.3,
+    convergence=DudsonConvergence(1e-4),
```

Comment:
Can change to 1e-6 (and will still work).

Reply:
done